features/environment.py

```diff
@@ -641,58 +658,21 @@ def create_instance_with_uat_installed(
             context.reuse_container[series],
         )
         return
-    ppa = context.config.ppa
-    if ppa == "":
-        image_name = context.config.cloud_manager.locate_image_name(series)
-        logging.info(
-            "--- Unset UACLIENT_BEHAVE_PPA. Using ubuntu-advantage-tools "
-            + "from image: {}".format(image_name)
-        )
-        context.series_image_name[series] = image_name
-        return
 
-    deb_paths = []
-    if context.config.build_pr:
-        deb_paths = build_debs_from_sbuild(context, series)
+    user_data = _get_user_data_for_instance(context.config, series)
 
     logging.info(
-        "--- Launching VM to create a base image with updated ubuntu-advantage"
+        "--- Launching VM to create a base image with ubuntu-advantage"
     )
-
-    user_data = ""
-    ud_file = context.config.userdata_file
-    if ud_file and os.path.exists(ud_file):
-        with open(ud_file, "r") as stream:
-            user_data = stream.read()
-    if "pro" in context.config.machine_type:
-        if not user_data:
-            user_data = USERDATA_BLOCK_AUTO_ATTACH_IMG
-    if not deb_paths:
-        if not user_data:
-            user_data = "#cloud-config\n"
-
-        if context.config.enable_proposed:
-            user_data += USERDATA_RUNCMD_ENABLE_PROPOSED.format(series=series)
-        else:
-            ppa = context.config.ppa
-            ppa_keyid = context.config.ppa_keyid
-            if context.config.ppa.startswith("ppa:"):
-                ppa = (
-                    ppa.replace("ppa:", "http://ppa.launchpad.net/")
-                    + "/ubuntu"
-                )
-
-            if ppa == DAILY_PPA:
-                user_data += USERDATA_RUNCMD_PIN_PPA
-
-            user_data += USERDATA_APT_SOURCE_PPA.format(
-                ppa_url=ppa, ppa_keyid=ppa_keyid
-            )
     inst = context.config.cloud_manager.launch(
         instance_name=name, series=series, user_data=user_data
     )
     instance_id = context.config.cloud_manager.get_instance_id(inst)
 
+    deb_paths = None
+    if context.config.install_from.value == InstallationSource.LOCAL.value:
```

Comment:
Do can't we just compare here without calling `value` ?

Reply:
Now this one is interesting - there is a known caveat for `Enum` comparison, due to how the import machinery works.
I found it weird because in our case there should be no `import`, the `Enum` subclass is defined in the same file where we use it. However, behave does some """trick""" in their runtime which screws this up.

Everything in `environment.py` gets to....... `builtins`. I am no Java specialist, so I don't know if it makes any sense. But take a look on why the direct comparison was failing (with a debugger set right after the config gets instantiated with `from_environ`):
```
(Pdb) InstallationSource.__module__
'builtins'
(Pdb) id(InstallationSource)
94603178161200
(Pdb) id(sys.modules["features.environment"].InstallationSource)
94603179143200
```
Python sees two different Enums with the same name. Now, when we are actually doing the comparison, the config value comes from the `builtins` Enum which behave creates, and the comparison is done against the original one:
```
(Pdb) InstallationSource.__module__
'features.environment'
(Pdb) config.install_from
<InstallationSource.DAILY: 'daily'>
(Pdb) id(config.install_from)
140671518494912
(Pdb) id(InstallationSource("daily"))
140671515232864
```

So the way I fixed it was cheating on how behave screws up the import machinery: I moved the Enum to the `utils` file. Since only the `environment` file "goes to `builtins`", behave is forced to import the Enum from utils when creating the config, and then we do the same when running the comparison. Back to config creation time, this time with the imported Enum:
```
(Pdb) InstallationSource.__module__
'features.util'
(Pdb) id(InstallationSource)
94407529466464
(Pdb) id(sys.modules["features.environment"].InstallationSource)
94407529466464
```

It made me question the structure a little: we could have only the "behave specific" functions (like `before_all` or `after_all` in the `environment.py` and have a separate `config.py` for our things... But to avoid the potential overkill, I just moved it to utils right now (:

Reply:
Wow, yeah I didn't know that about `environment.py`. I agree that in some future PR, we should refcator so that only behave specific functionality that should be treated this way is in that file, and everything else can get factored out.